src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,278 @@
+
+  public StarlarkWasmModule(
+      StarlarkPath path,
+      Object origPath,
+      byte[] moduleContent,
+      String allocateFn)
+      throws EvalException {
+    WasmModule wasmModule;
+    Profiler prof = Profiler.instance();
+    try (SilentCloseable c1 = prof.profile(WASM_LOAD, () -> "load " + path.toString())) {
+      try (SilentCloseable c2 = prof.profile(WASM_LOAD, "parse")) {
+        try {
+          wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+        } catch (ChicoryException e) {
+          throw new EvalException(e);
+        }
+      }
+      validateModule(wasmModule, allocateFn);
+    }
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    Profiler prof = Profiler.instance();
+    try (SilentCloseable c = prof.profile(WASM_EXEC, () -> "execute " + functionName)) {
+      var memLimits = getMemLimits(memLimitBytes);
+      // Perform initialization and execution in a separate thread so it can be interrupted
+      // in case of timeout.
+      var wasmThreadFactory =
+          Thread.ofPlatform().name(Thread.currentThread().getName() + "_wasm").factory();
+      StarlarkWasmExecutionResult result;
+      String errMessage;
+      try (var executor = Executors.newSingleThreadExecutor(wasmThreadFactory)) {
+        return executor.invokeAny(
+            ImmutableList.of(() -> run(functionName, input, memLimits)),
+            timeout.toMillis(),
+            TimeUnit.MILLISECONDS);
+      } catch (TimeoutException e) {
+        errMessage = String.format("Error executing %s: timed out", functionName);
+      } catch (ExecutionException e) {
+        errMessage = String.format("Error executing %s: %s", functionName, e.getCause().getMessage());
+      }
+      return StarlarkWasmExecutionResult.newErr(errMessage);
+    }
+  }
+
+  private StarlarkWasmExecutionResult run(String execFunc, byte[] input, MemoryLimits memLimits)
+      throws EvalException, InterruptedException {
+    Instance instance;
+    Profiler prof = Profiler.instance();
+    try {
+      instance = Instance.builder(wasmModule)
+          .withMemoryLimits(memLimits)
+          // Disable calling `_start()`, which is the entry point for WASI-style
```

Comment:
this doesn't seem accurate. Per [chicory source code](https://github.com/dylibso/chicory/blob/f52853d37628c719efd4027a4b76c708cb5dd4df/runtime/src/main/java/com/dylibso/chicory/runtime/Instance.java#L789) and the [WASM spec](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-start), this is just the function to be run when the module is instantiated. Is it safe to disable it here? (WASI doesn't seem like a concern since we're not using the WASI part of chicory at all.)

Reply:
I expanded the nearby comment. I agree with you that the plain language of the spec says that the initialization function should be named in the start section (and probably named `_start`?), but actual implementations use the start section for the entry point (like C's `main`) and put early initialization into `_initialize`.

There's a lot of discussion about this at the following threads if you're looking for some light bedtime reading:
* https://bugs.llvm.org/show_bug.cgi?id=37198
* https://reviews.llvm.org/D40559
* https://github.com/WebAssembly/design/issues/1160

Reply:
I haven't read all of this (it's not bedtime yet), but comments such as https://github.com/WebAssembly/design/issues/1160#issuecomment-493930514 seem to support Xudong's understanding. Are we certain that every reasonable implementation shares the same "misreading" of the spec or could we end up in a split-world situation where some implementations work and others rightly complain about us failing to adhere to the spec?

Reply:
thanks for the context. I also haven't read through everything, but I'll take your word for it for now. Could you pelase add these links to the code comment for posterity?

Reply:
Added the above links to the comment.